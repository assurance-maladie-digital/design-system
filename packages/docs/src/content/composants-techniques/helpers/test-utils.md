---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
