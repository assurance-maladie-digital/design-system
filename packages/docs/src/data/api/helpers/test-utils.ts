import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
