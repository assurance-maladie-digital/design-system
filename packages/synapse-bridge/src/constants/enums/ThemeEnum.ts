export enum ThemeEnum {
	AMELI = 'ameli',
	AMELI_PRO = 'ameli-pro',
	CNAM = 'cnam',
	COMPTE_AMELI = 'compte-ameli',
	COMPTE_ENTREPRISE = 'compte-entreprise',
	DEFAULT = 'default',
	RISQUE_PRO = 'risque-pro'
}

export const THEME_ENUM_VALUES = Object.values(ThemeEnum);
