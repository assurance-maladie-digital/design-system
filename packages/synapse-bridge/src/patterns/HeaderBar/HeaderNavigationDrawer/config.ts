export const config = {
	app: true,
	dark: true,
	temporary: true,
	width: '320px',
	class: 'pa-4',
	closeBtn: {
		icon: true
	},
	tabs: {
		dense: true,
		vertical: true,
		backgroundColor: 'transparent'
	},
	tab: {
		exact: true,
		class: 'justify-start mb-2'
	}
};
