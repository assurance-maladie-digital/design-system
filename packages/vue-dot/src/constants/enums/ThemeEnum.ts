export enum ThemeEnum {
	DEFAULT = 'default',
	CNAM = 'cnam',
	AMELI = 'ameli',
	AMELI_PRO = 'ameli-pro',
	COMPTE_AMELI = 'compte-ameli',
	RISQUE_PRO = 'risque-pro',
	COMPTE_ENTREPRISE = 'compte-entreprise'
}

export const THEME_ENUM_VALUES = Object.values(ThemeEnum);
