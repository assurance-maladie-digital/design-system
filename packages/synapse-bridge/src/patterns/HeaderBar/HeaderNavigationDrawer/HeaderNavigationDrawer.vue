<script lang="ts">
import { defineComponent } from "vue";
import type { PropType } from "vue";
import { mdiClose } from "@mdi/js";
import { config } from "./config";
import { locales } from "./locales";
import { NavigationItem } from "../types";
import { colorMapping } from "../colorMapping";
import { ThemeEnum } from "@/constants/enums/ThemeEnum";
import { customizable } from "@/mixins/customizable";
export default defineComponent({
	mixins: [customizable(config)],
	props: {
		theme: {
			type: String as PropType<ThemeEnum>,
			required: true,
		},
		items: {
			type: Array as PropType<NavigationItem[]>,
			default: undefined,
		},
		mobileVersion: {
			type: Boolean,
			default: false,
		},
		drawer: {
			type: Boolean,
			default: false,
		},
		tab: {
			type: [Number, String],
			default: null,
		},
	},
	emits: ["change", "update:tab"],
	data() {
		return {
			closeIcon: mdiClose,
			locales,
		};
	},
	computed: {
		spacingClass(): string {
			return this.mobileVersion ? 'px-4' : 'px-14';
		},
		backgroundColor(): string {
			return colorMapping[this.theme];
		}
	},
	methods: {
		drawerUpdated(value: boolean): void {
			if (!value) {
				this.$emit('change', false);
			}
		},
		emitChangeEvent(): void {
			this.$emit('change', !this.drawer);
		},
		emitTabUpdateEvent(value: number): void {
			this.$emit('update:tab', value);
		}
	}
});
</script>

<template>
	<VNavigationDrawer
		v-if="mobileVersion && items"
		:color="backgroundColor"
		:modelValue="drawer"
		v-bind="options"
		@input="drawerUpdated"
	>
		<slot>
			<div class="d-flex align-center justify-end mb-8">
				<VBtn
					v-bind="options.closeBtn"
					variant="text"
					:aria-label="locales.closeBtn"
					@click="emitChangeEvent"
				>
					<VIcon v-bind="options.closeIcon">
						{{ closeIcon }}
					</VIcon>
				</VBtn>
			</div>

			<slot name="navigation-drawer-prepend" />

			<VTabs
				v-bind="options.tabs"
				:modelValue="tab"
				class="flex-grow-0"
				@change="emitTabUpdateEvent"
			>
				<VTab
					v-for="(item, index) in items"
					:key="index"
					:href="item.href"
					:to="item.to"
					v-bind="options.tab"
				>
					{{ item.label }}
				</VTab>
			</VTabs>

			<slot name="navigation-drawer-append" />
		</slot>
	</VNavigationDrawer>
</template>

<style lang="scss" scoped>
.v-navigation-drawer {
	:deep(.v-navigation-drawer__content) {
		display: flex;
		flex-direction: column;
		align-items: flex-start;
		.v-slide-group {
			width: 100%;
		}
		.v-tab {
			text-transform: uppercase;
			letter-spacing: .0892857143em;
		}
		.v-tab--selected {
			background: rgba($color: #fff, $alpha: 0.1);
		}
	}
}
</style>
