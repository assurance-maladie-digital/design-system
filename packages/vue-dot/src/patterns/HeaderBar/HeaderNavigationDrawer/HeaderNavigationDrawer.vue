<template>
	<VNavigationDrawer
		v-if="mobileVersion && items"
		v-bind="options"
		:value="drawer"
		:color="backgroundColor"
		@input="drawerUpdated"
	>
		<slot>
			<div class="d-flex align-center justify-end mb-8">
				<VBtn
					v-bind="options.closeBtn"
					:aria-label="locales.closeBtn"
					@click="emitChangeEvent"
				>
					<VIcon v-bind="options.closeIcon">
						{{ closeIcon }}
					</VIcon>
				</VBtn>
			</div>

			<slot name="navigation-drawer-prepend" />

			<VTabs
				v-bind="options.tabs"
				:value="tab"
				optional
				class="flex-grow-0"
				@change="emitTabUpdateEvent"
			>
				<VTab
					v-for="(item, index) in items"
					:key="index"
					v-bind="options.tab"
					:href="item.href"
					:to="item.to"
				>
					{{ item.label }}
				</VTab>
			</VTabs>

			<slot name="navigation-drawer-append" />
		</slot>
	</VNavigationDrawer>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { mdiClose } from '@mdi/js';

	import { config } from './config';
	import { locales } from './locales';
	import { NavigationItem } from '../types';
	import { colorMapping } from '../colorMapping';

	import { ThemeEnum } from '../ThemeEnum';

	import { customizable } from '../../../mixins/customizable';

	const Props = Vue.extend({
		props: {
			theme: {
				type: String as PropType<ThemeEnum>,
				required: true
			},
			items: {
				type: Array as PropType<NavigationItem[]>,
				default: undefined
			},
			mobileVersion: {
				type: Boolean,
				default: false
			},
			drawer: {
				type: Boolean,
				default: false
			},
			tab: {
				type: [Number, String],
				default: null
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config));

	@Component({
		model: {
			prop: 'drawer',
			event: 'change'
		}
	})
	export default class HeaderNavigationDrawer extends MixinsDeclaration {
		locales = locales;

		closeIcon = mdiClose;

		get spacingClass(): string {
			return this.mobileVersion ? 'px-4' : 'px-14';
		}

		get backgroundColor(): string {
			return colorMapping[this.theme];
		}

		drawerUpdated(value: boolean): void {
			if (value === false) {
				this.$emit('change', false);
			}
		}

		emitChangeEvent(): void {
			this.$emit('change', !this.drawer);
		}

		emitTabUpdateEvent(value: number): void {
			this.$emit('update:tab', value);
		}
	}
</script>

<style lang="scss" scoped>
	.v-navigation-drawer {
		.v-tab {
			height: 40px !important;

			&.v-tab--active {
				background: rgba($color: #fff, $alpha: .1);
			}
		}

		:deep(.v-navigation-drawer__content) {
			display: flex;
			flex-direction: column;
			align-items: flex-start;
		}
	}
</style>
