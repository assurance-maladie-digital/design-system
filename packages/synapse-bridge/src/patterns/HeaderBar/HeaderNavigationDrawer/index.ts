import HeaderNavigationDrawer from './HeaderNavigationDrawer.vue';

export default HeaderNavigationDrawer;
