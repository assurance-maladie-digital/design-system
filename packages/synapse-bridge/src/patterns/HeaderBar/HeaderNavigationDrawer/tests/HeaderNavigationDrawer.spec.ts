import { describe, it, expect } from "vitest";
import { shallowMount } from "@vue/test-utils";
import { vuetify } from "@tests/unit/setup";

import HeaderNavigationDrawer from "../";
import { ThemeEnum } from "@/constants/enums/ThemeEnum";

describe("HeaderNavigationDrawer", () => {
	it("renders correctly", () => {
		const wrapper = shallowMount(HeaderNavigationDrawer, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				theme: ThemeEnum.DEFAULT,
				mobileVersion: true,
			},
		});

		expect(wrapper).toMatchSnapshot();
	});

	it("emits change event when drawer is toggled", async () => {
		const wrapper = shallowMount(HeaderNavigationDrawer, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				theme: ThemeEnum.DEFAULT,
				mobileVersion: true,
			},
		});

		wrapper.vm.emitChangeEvent();
		expect(wrapper.emitted("change")).toBeTruthy();
		expect(wrapper.emitted("change")).toHaveLength(1);
	});

	it("emits tab update event when tab is changed", async () => {
		const items = [
			{ label: "Home", href: "#home" },
			{ label: "About", href: "#about" },
		];

		const wrapper = shallowMount(HeaderNavigationDrawer, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				theme: ThemeEnum.DEFAULT,
				mobileVersion: true,
				items: items,
			},
		});

		wrapper.vm.emitTabUpdateEvent(1);
		expect(wrapper.emitted("update:tab")).toBeTruthy();
		expect(wrapper.emitted("update:tab")).toHaveLength(1);
	});

	it("returns correct spacing class based on mobileVersion prop", () => {
		const wrapperMobile = shallowMount(HeaderNavigationDrawer, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				theme: ThemeEnum.DEFAULT,
				mobileVersion: true,
			},
		});

		expect(wrapperMobile.vm.spacingClass).toBe("px-4");

		const wrapperDesktop = shallowMount(HeaderNavigationDrawer, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				theme: ThemeEnum.DEFAULT,
				mobileVersion: false,
			},
		});
		expect(wrapperDesktop.vm.spacingClass).toBe("px-14");
	});

	it("emits change event with false when drawerUpdated is called with value false", async () => {
		const wrapper = shallowMount(HeaderNavigationDrawer, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				theme: ThemeEnum.DEFAULT,
				mobileVersion: true,
			},
		});

		wrapper.vm.drawerUpdated(false);

		await wrapper.vm.$nextTick();

		expect(wrapper.emitted("change")).toBeTruthy();
		expect(wrapper.emitted("change")).toHaveLength(1);
	});

	it("doesn't emit change event when drawerUpdated is called with value true", async () => {
		const wrapper = shallowMount(HeaderNavigationDrawer, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				theme: ThemeEnum.DEFAULT,
				mobileVersion: true,
			},
		});

		wrapper.vm.drawerUpdated(true);

		await wrapper.vm.$nextTick();

		expect(wrapper.emitted("change")).toBeFalsy();
	});
});
