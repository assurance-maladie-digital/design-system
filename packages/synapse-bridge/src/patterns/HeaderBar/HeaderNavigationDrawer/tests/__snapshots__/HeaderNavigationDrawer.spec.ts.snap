// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeaderNavigationDrawer > renders correctly 1`] = `<!--v-if-->`;
